### PR TITLE
fix: [CI-9556]: Add support for log-service acl check

### PR DESCRIPTION
**Pull request by smjt-h**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/55909)

### CODE REVIEW

**Comment by smjt-h** [comment]
trigger uts

**Comment by smjt-h** [comment]
trigger uts

**Comment by smjt-h** [comment]
trigger uts

**Comment by smjt-h** [comment]
trigger uts

File: product/log-service/handler/middleware.go [hunk 1]
```diff
@@ -155,6 +160,76 @@ func AuthMiddleware(config config.Config, ngClient *client.HTTPClient, skipKeyCh
...
+			//then check acl with call to access control, else check for old approach (x-api-key or X-harness-Token)
+			if inputAuthToken != "" && r.Method == http.MethodGet {
+				if r.FormValue(pipelineIdentifier) == "" || r.FormValue(projectIdentifier) == "" || r.FormValue(orgIdentifier) == "" {
+					WriteBadRequest(w, errors.New("Scope pipelineID, projectID and orgID are required for validating access"))
+					return
+				}
+
+				allowed, err := aclClient.ValidateAccessforPipeline(r.Context(), inputAuthToken, accountID, r.FormValue(pipelineIdentifier), r.FormValue(projectIdentifier), r.FormValue(orgIdentifier))
+				if err != nil {
+					fmt.Println("error Validating Access to pipeline: ", pipelineIdentifier, err)
```
Comment (vistaarjuneja) [inline product/log-service/handler/middleware.go:188]: Use logger instead of fmt everywhere

File: product/log-service/handler/middleware.go [hunk 2]
```diff
@@ -155,6 +160,76 @@ func AuthMiddleware(config config.Config, ngClient *client.HTTPClient, skipKeyCh
...
+					WriteBadRequest(w, errors.New("Error validating access for resource, unauthorized or expired token"))
```
Comment (vistaarjuneja) [inline product/log-service/handler/middleware.go:189]: This should be an internal error, not a bad request

File: product/log-service/handler/middleware.go [hunk 3]
```diff
@@ -155,6 +160,76 @@ func AuthMiddleware(config config.Config, ngClient *client.HTTPClient, skipKeyCh
...
+					return
+				}
+				if !allowed {
+					WriteBadRequest(w, errors.New("User not authorized"))
```
Comment (vistaarjuneja) [inline product/log-service/handler/middleware.go:193]: This should be a 403 access denied error

File: product/log-service/handler/middleware.go [hunk 4]
```diff
@@ -155,6 +160,76 @@ func AuthMiddleware(config config.Config, ngClient *client.HTTPClient, skipKeyCh
...
+					return
+				}
+			} else if inputApiKey != "" {
+				// Try to check token from the header or the URL param
+				err := doApiKeyAuthentication(inputApiKey, r.FormValue(accountIDParam), r.FormValue(routingIDparam), ngClient)
+				if err != nil {
+					WriteBadRequest(w, errors.New("apikey in request not authorized for receiving tokens"))
```
Comment (vistaarjuneja) [inline product/log-service/handler/middleware.go:200]: This should be internal error - have a look at all returns and update to an error code which makes sense

File: product/log-service/handler/middleware.go [hunk 5]
```diff
@@ -155,6 +160,76 @@ func AuthMiddleware(config config.Config, ngClient *client.HTTPClient, skipKeyCh
...
+
+				if inputToken == "" {
+					WriteBadRequest(w, errors.New("no token in header"))
+					return
+				}
+				// accountID in token should be same as accountID in URL
+				secret := []byte(config.Auth.LogSecret)
+				login := authcookie.Login(inputToken, secret)
+				if login == "" || login != accountID {
+					WriteBadRequest(w, errors.New(fmt.Sprintf("operation not permitted for accountID: %s", accountID)))
```
Comment (vistaarjuneja) [inline product/log-service/handler/middleware.go:217]: 403 access denied error - we should fix it in other places as well if we have it other than this PR

File: product/platform/client/helper.go [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+package client
+
+const (
+	resource_pipeline = "PIPELINE"
+	pipeline_view     = "core_pipeline_view"
+)
+
+// getACLRequest returns an ACLRequest object with the specified account ID, pipeline ID, project ID, and organization ID.
+func GetACLRequest(accountID, pipelineID, projectID, orgID string) ACLRequest {
```
Comment (vistaarjuneja) [inline product/platform/client/helper.go:14]: A function called `GetACLRequest` in `platform/client/helper` shouldn't be hardcoding the resource permission and resource type.
If it sits in `product/platform`, the client should be generic and not be tied to the code needed for log service

File: product/log-service/handler/handler.go [hunk 7]
```diff
@@ -81,7 +81,7 @@ func Handler(queue queue.Queue, cache cache.Cache, stream stream.Stream, store s
 		sr := chi.NewRouter()
 		// Validate the accountID in URL with the token generated above and authorize the request
 		if !config.Auth.DisableAuth {
-			sr.Use(AuthMiddleware(config, ngClient, false))
```
Comment (vistaarjuneja) [inline product/log-service/handler/handler.go:84]: We can keep the name AuthMiddleware and just update that handler directly

**Review by vistaarjuneja — CHANGES_REQUESTED** [review]
left comments

File: product/log-service/config/config.go [hunk 8]
```diff
@@ -71,7 +72,7 @@ type Config struct {
 	}
 
 	ConsumerWorker struct {
-		WorkerPool    int    `envconfig:"LOG_SERVICE_WORKER_POOL" default:"1"`
+		WorkerPool    int    `envconfig:"LOG_SERVICE_WORKER_POOL" default:"0"`
```
[the remark this answers is not in the document]
Comment (smjt-h) [reply product/log-service/config/config.go:75]: reverted

File: product/platform/client/client.go [hunk 9]
```diff
@@ -92,8 +93,51 @@ type Account struct {
...
+	AccountIdentifier string `json:"accountIdentifier"`
+	OrgIdentifier     string `json:"orgIdentifier"`
+	ProjectIdentifier string `json:"projectIdentifier"`
+}
+
+// AccountResponse represents a json-encoded response for /acl api.
+type ACLResponse struct {
+	Status string `json:"status"`
+	Data   struct {
+		Principal struct {
```
[the remark this answers is not in the document]
Comment (smjt-h) [reply product/platform/client/client.go:118]: done

File: product/platform/client/client.go [hunk 10]
```diff
@@ -92,8 +93,51 @@ type Account struct {
 	ResponseMessages []interface{} `json:"responseMessages"`
 }
 
+// AccountResponse represents a json-encoded response for /acl api.
```
[the remark this answers is not in the document]
Comment (smjt-h) [reply product/platform/client/client.go:96]: done

File: product/log-service/handler/middleware.go (hunk 1 above)
Comment (smjt-h) [reply product/log-service/handler/middleware.go:188]: done

File: product/log-service/handler/middleware.go (hunk 2 above)
Comment (smjt-h) [reply product/log-service/handler/middleware.go:189]: done

File: product/log-service/handler/middleware.go (hunk 3 above)
Comment (smjt-h) [reply product/log-service/handler/middleware.go:193]: done

File: product/log-service/handler/middleware.go (hunk 4 above)
Comment (smjt-h) [reply product/log-service/handler/middleware.go:200]: fixed

File: product/log-service/handler/middleware.go (hunk 5 above)
Comment (smjt-h) [reply product/log-service/handler/middleware.go:217]: fixed

File: product/log-service/handler/handler.go (hunk 7 above)
Comment (smjt-h) [reply product/log-service/handler/handler.go:84]: done

File: product/platform/client/client.go [hunk 11]
```diff
@@ -92,8 +93,62 @@ type Account struct {
...
+	Permission         string        `json:"permission"`
+}
+
+type ResourceScope struct {
+	AccountIdentifier string `json:"accountIdentifier"`
+	OrgIdentifier     string `json:"orgIdentifier"`
+	ProjectIdentifier string `json:"projectIdentifier"`
+}
+
+// ACLPrincipal represents the principal information.
```
Comment (vistaarjuneja) [inline product/platform/client/client.go:114]: these comments are not useful, you can remove unless there's some net new information. Or we take the approach that we always add comments everywhere (which we don't)

File: product/log-service/handler/middleware.go [hunk 12]
```diff
@@ -117,12 +123,35 @@ func AuthMiddleware(config config.Config, ngClient *client.HTTPClient, skipKeyCh
...
 			}
 
-			// Try to get token from the header or the URL param
 			inputApiKey := r.Header.Get(authAPIKeyHeader)
-			if inputApiKey != "" {
+			inputAuthToken := r.Header.Get(authTokenHeader)
+
+			//Check if Auth token is present in Header and if method is allowed,
+			//then check acl with call to access control, else check for old approach (x-api-key or X-harness-Token)
+			if inputAuthToken != "" && r.Method == http.MethodGet {
```
Comment (vistaarjuneja) [inline product/log-service/handler/middleware.go:131]: Why are we checking that the method is GET? Shouldn't we be doing authorization for all the API calls?

File: product/log-service/handler/middleware.go [hunk 13]
```diff
@@ -117,12 +123,35 @@ func AuthMiddleware(config config.Config, ngClient *client.HTTPClient, skipKeyCh
...
+				if err != nil {
+					logger.FromRequest(r).
+						WithError(err).
+						WithField("pipelineIdentifier", pipelineIdentifier).
+						Errorln("error Validating Access to pipeline:")
+					WriteInternalError(w, errors.New("Error validating access for resource, unauthorized or expired token"))
+					return
+				}
+				if !allowed {
+					writeError(w, errors.New("User not authorized"), 403)
```
Comment (vistaarjuneja) [inline product/log-service/handler/middleware.go:147]: lower case

File: product/log-service/handler/middleware.go (hunk 12 above)
Comment (smjt-h) [reply product/log-service/handler/middleware.go:131]: We dont want to expose apis for all the calls for now we decided only for GET calls

File: product/platform/client/helper.go [hunk 14]
```diff
@@ -0,0 +1,28 @@
+// Copyright 2023 Harness Inc. All rights reserved.
+// Use of this source code is governed by the PolyForm Free Trial 1.0.0 license
+// that can be found in the licenses directory at the root of this repository, also available at
+// https://polyformproject.org/wp-content/uploads/2020/05/PolyForm-Free-Trial-1.0.0.txt.
+
+package client
+
+const (
+	resource_pipeline = "PIPELINE"
```
[the remark this answers is not in the document]
Comment (smjt-h) [reply product/platform/client/helper.go:9]: done

File: product/log-service/handler/middleware.go [hunk 15]
```diff
@@ -117,12 +123,35 @@ func AuthMiddleware(config config.Config, ngClient *client.HTTPClient, skipKeyCh
...
+					WriteBadRequest(w, errors.New("Scope pipelineID, projectID and orgID are required for validating access"))
+					return
+				}
+
+				allowed, err := aclClient.ValidateAccessforPipeline(r.Context(), inputAuthToken, accountID, r.FormValue(pipelineIdentifier), r.FormValue(projectIdentifier), r.FormValue(orgIdentifier))
+				if err != nil {
+					logger.FromRequest(r).
+						WithError(err).
+						WithField("pipelineIdentifier", pipelineIdentifier).
+						Errorln("error Validating Access to pipeline:")
```
[the remark this answers is not in the document]
Comment (smjt-h) [reply product/log-service/handler/middleware.go:142]: done

File: product/log-service/handler/middleware.go [hunk 16]
```diff
@@ -117,12 +123,35 @@ func AuthMiddleware(config config.Config, ngClient *client.HTTPClient, skipKeyCh
...
-			// Try to get token from the header or the URL param
 			inputApiKey := r.Header.Get(authAPIKeyHeader)
-			if inputApiKey != "" {
+			inputAuthToken := r.Header.Get(authTokenHeader)
+
+			//Check if Auth token is present in Header and if method is allowed,
+			//then check acl with call to access control, else check for old approach (x-api-key or X-harness-Token)
+			if inputAuthToken != "" && r.Method == http.MethodGet {
+				if r.FormValue(pipelineIdentifier) == "" || r.FormValue(projectIdentifier) == "" || r.FormValue(orgIdentifier) == "" {
+					WriteBadRequest(w, errors.New("Scope pipelineID, projectID and orgID are required for validating access"))
```
[the remark this answers is not in the document]
Comment (smjt-h) [reply product/log-service/handler/middleware.go:133]: done

**Comment by smjt-h** [comment]
trigger uts

**Review by vistaarjuneja — APPROVED** [review]

**Pull request merged**